mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td

```diff
@@ -260,74 +256,380 @@ def MIOpen_WorkitemIdOp:
+      iter_args(%arg0 = %c0 : index) bounds [4, 8] {
+      %v = miopen.buffer_load {...} %place[%a1, %a2, %a3]
+      miopen.buffer_store {...} %v -> %buffer[%i, %j]
+      %cont = addi %arg0, %c1
+      miopen.yield %cont : index
+    }
+    ```
+    will lower to
+    ```mlir
+    %a0, %a1, %a2 = miopen.compute_transforms [#transform_map1](%x, %y)
+    %i, %j = miopen.compute_transforms [](%c0, %c0)
+    %a0_0_0, %a1_0_0 = miopen.index_diff #transform_map1[0, 0] (%x, %y) -> (%a0, %a1, %a2)
+    %i_0_0 = addi %i, %cst0
+    %j_0_0 = addi %j, %cst0
+    %v_0_0 miopen.buffer_load {...} %place[%a0_0_0, %a1_0_0, %a2_0_0]
+    miopen.buffer_store {...} %v -> %buffer[%i_0_0, %j_0_0]
+    %res_0_0 = addi %c0, %c1 : inedx
+    %a0_0_1, %a1_0_1, %a2_0_1 = miopen.index_diff #transform_map1[0, 1] (%x, %y) -> (%a0, %a1, %a2)
+    ...
+    %res_3_7 = addi %res_3_6, %c1 : index
+  ```
+  Or, if no unrolling or index diffs are called for
+  ```mlir
+  %res = affine for %d0 = 0 to 4 iter_args(%arg0 = %c0) {
+    %res0 = affine.for %d1 = 0 to 8 iter_args(%arg1 = %arg0) {
+      %upper_0_0 = addi %x, %d0
+      %upper_0_1 = addi %y, %d1
+      %i = addi %c0, %d0
+      %j = addi %c0, %d1
+      %a0, %a1, %a2 = affine.apply (#transform_map1.map) (%upper_0_0, %upper_0_1)
+      %v = miopen.buffer_load {...} %place[%a0, %a1, %a2]
+      miopen.buffer_store {...} %buffer[%i, %j]
+      %cont = addi %arg1, %c1 : index
+      affine.yield %cont : index
+    }
+    affine.yield %res0 : index
+  }
+  ```
+
+  The purpose of this looping construct is to remove redundancy and make the
+  application of coordinate transformations more explicit in the IR.
+  }];
+
+  let regions = (region SizedRegion<1>:$region);
+
+  DerivedAttr upperLen = DerivedAttr<"uint32_t", [{
+    return upperInits().size() / transforms().size();
+  }], [{ $_builder.getI32IntegerAttr($_self) }]>;
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    OpBuilder<(ins "ArrayRef<ValueRange>":$inits,
+    "ArrayRef<Attribute>":$transforms,
+    "ArrayRef<int64_t>":$bounds, "bool":$forceUnroll, "bool":$useIndexDiffs,
+    CArg<"ValueRange", "llvm::None">:$iterArgs)>,
+
+    OpBuilder<(ins "ArrayRef<ValueRange>":$inits,
+    "ArrayRef<Attribute>":$transforms,
+    "ArrayAttr":$bounds, "bool":$forceUnroll, "bool":$useIndexDiffs,
+    CArg<"ValueRange", "llvm::None">:$iterArgs)>,
+
+    OpBuilder<(ins "ArrayRef<ValueRange>":$inits,
+    "ArrayAttr":$transforms,
+    "ArrayRef<int64_t>":$bounds, "bool":$forceUnroll, "bool":$useIndexDiffs,
+    CArg<"ValueRange", "llvm::None">:$iterArgs)>,
+
+    OpBuilder<(ins "ArrayRef<ValueRange>":$inits,
+    "ArrayAttr":$transforms,
+    "ArrayAttr":$bounds, "bool":$forceUnroll, "bool":$useIndexDiffs,
+    CArg<"ValueRange", "llvm::None">:$iterArgs)>];
+
+  let extraClassDeclaration = [{
+    uint32_t lowerStart(uint32_t n) {
+      return *(lowerStarts().getValues<uint32_t>().begin() + n);
+    }
+
+    Block::BlockArgListType getLowerCoords(uint32_t domain) {
```

Comment:
From the call sites it appears this method is crucial so documentation is needed. It's not entirely obvious what each domain is for. And the concept of `domain` is not defined in the documentation of `transforming_for`.